cmd/snap-update-ns/utils.go

```diff
@@ -87,6 +89,46 @@ func IsReadOnly(dirFd int, dirName string) (bool, error) {
 	return false, nil
 }
 
+// IsSnapdCreatedPrivateTmpfs returns true if a directory is a tmpfs mounted by snapd.
+//
+// The function inspects the directory (represented as both an open file
+// descriptor and the absolute path) and a list of changes that were applied to
+// the mount namespace. A directory is trusted if it is a tmpfs that was
+// mounted by snap-confine or snapd-update-ns. Note that sub-directories of a
+// trusted tmpfs are not considered trusted by this function.
+func IsSnapdCreatedPrivateTmpfs(dirFd int, dirName string, changes []*Change) (bool, error) {
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(dirFd, &fsData); err != nil {
+		return false, fmt.Errorf("cannot fstatfs %q: %s", dirName, err)
+	}
+	// If something is not a tmpfs it cannot be the trusted tmpfs we are looking for.
+	if fsData.Type != TmpfsMagic {
+		return false, nil
+	}
+	// Any of the past changes that mounted a tmpfs exactly at the directory we
+	// are inspecting is considered as trusted. This is conservative because it
+	// doesn't trust sub-directories of a trusted tmpfs. This approach is
+	// sufficient for the intended use.
+	//
+	// The algorithm goes over all the changes in reverse and picks up the
+	// first tmpfs mount or unmount action that matches the directory name.
+	// The set of constraints in snap-update-ns and snapd prevent from mounting
+	// over an existing mount point so we don't need to consider e.g. a bind
+	// mount shadowing an active tmpfs.
+	for i := len(changes) - 1; i >= 0; i-- {
+		change := changes[i]
+		if change.Entry.Type == "tmpfs" && change.Entry.Dir == dirName {
+			return change.Action == Mount, nil
```

Comment:
This doesn't match the above comment ("...picks up the first tmpfs mount or unmount action..."). Did you forget about Unmount here, or is the comment incorrect?

Reply:
No, the comment is correct. Since we only have Mount and Unmount actions we look for the first change that _either_ mounts or unmounts a tmpfs and then return our answer. The boolean flag in our answer depends on the action, for Mount we return true, for Unmount we return false.

Reply:
I see, makes sense, thanks!